src/vmm/src/lib.rs

```diff
@@ -355,25 +363,12 @@ impl Vmm {
             .map_err(Error::I8042Error)
     }
 
-    /// Waits for all vCPUs to exit and terminates the Firecracker process.
-    pub fn stop(&mut self, exit_code: i32) {
+    /// Signals Vmm should exit.
+    pub fn stop(&mut self, exit_code: ExitCode) {
         info!("Vmm is stopping.");
-
-        if let Some(observer) = self.events_observer.as_mut() {
-            if let Err(e) = observer.on_vmm_stop() {
-                warn!("{}", Error::VmmObserverTeardown(e));
-            }
-        }
-
-        // Write the metrics before exiting.
-        if let Err(e) = METRICS.write() {
-            error!("Failed to write metrics while stopping: {}", e);
-        }
-
-        // Exit from Firecracker using the provided exit code. Safe because we're terminating
-        // the process anyway.
-        unsafe {
-            libc::_exit(exit_code);
+        self.shutdown_exit_code = Some(exit_code);
+        if let Err(e) = self.exit_evt.write(1) {
```

Comment:
If I understand correctly, this function is called when the `exit_evt` was signaled. And in this same function, we signal `exit_evt` again and set the exit code, which seems a little redundant to me. Can't we do all of this in the Subscriber::process implementation?

Is this a workaround to the fact that there is currently no way of breaking out of the event manager loop?

Reply:
This is part of the mess/cyclic code that was ultimately refactored. The `Vcpus` commit will clean up and document the teardown protocol with its communication and steps. Please take a look at the final code.

> I am wondering whether there is significant impact on the shutdown time. Do you view this as critical/something to consider?

I may be experiencing a failure of imagination, but I can't think of any production scenario where process closure time would be so critical/latency-sensitive as to be impacted by graceful teardown.

Reply:
> This is part of the mess/cyclic code that was ultimately refactored. The Vcpus commit in this PR cleans up and documents the teardown protocol with its communication and steps. Please take a look at the final code.

Cool, I usually review code commit-by-commit :D

Reply:
> I may be experiencing a failure of imagination, but I can't think of any production scenario where process closure time would be so critical/latency-sensitive as to be impacted by graceful teardown.

I think you're right. Anyone who wants to quickly kill the Firecracker process can just send a SIGKILL.